src/server/ua_services_nodemanagement.c

```diff
@@ -924,6 +924,80 @@ recursiveTypeCheckAddChildren(UA_Server *server, UA_Session *session,
     return UA_STATUSCODE_GOOD;
 }
 
+/* Check if we got a valid browse name for the new node.
+ * For object nodes the BrowseName may only be null if the parent type has a
+ * 'DefaultInstanceBrowseName' property.
+ * */
+static UA_StatusCode
+checkValidBrowseName(UA_Server *server, UA_Session *session,
+							  const UA_Node **nodeptr, const UA_Node *type) {
+
+	UA_assert(type != NULL);
+	UA_StatusCode retval = UA_STATUSCODE_GOOD;
+	const UA_Node *node = *nodeptr;
+
+	if (node->nodeClass != UA_NODECLASS_OBJECT) {
+		if (UA_QualifiedName_isNull(&node->browseName))
+			return UA_STATUSCODE_BADBROWSENAMEINVALID;
+		return UA_STATUSCODE_GOOD;
+	}
+
+	if (!UA_QualifiedName_isNull(&node->browseName))
+		return UA_STATUSCODE_GOOD;
+
+	/* at this point we have an object with an empty browse name.
+	 * Check the type node if it has a DefaultInstanceBrowseName property
+	 */
+
+	UA_Variant defaultBrowseName;
+	UA_Variant_init(&defaultBrowseName);
+
+	UA_NodeId hasPropertyReference = UA_NODEID_NUMERIC(0, UA_NS0ID_HASPROPERTY);
```

Comment:
Use the TranslateBrowsePathToNodeId service to get a list of candidates.

Browse/View should be implemented by just a few methods that are reused.
So we can change node internals later on with minimal impact.